### PR TITLE
Ensemble!

### PR DESCRIPTION
**Pull request by jmccreight**
All new add, compose, and run methods. 
No scheduler work yet. 
All tested!

### CODE REVIEW

**Comment by jmills-ncar** [comment]
great coverage, nice job!

File: wrfhydropy/core/ensemble.py [hunk 1]
```diff
@@ -0,0 +1,369 @@
...
+
+        if len(self) == 1:
+            print('Ensemble is of length 1, no differences.')
+            return {}
+
+        mem_0_ref_dict = dictify(self.members[0])
+
+        # TODO(JLM): Could this be parallelized?
+        all_diff_keys = set({})
+        for ii in range(1, len(self)):
```
Comment (jmills-ncar) [inline wrfhydropy/core/ensemble.py]: I belive `enumerate()` is preferred over `range()`

File: wrfhydropy/core/ensemble.py [hunk 2]
```diff
@@ -0,0 +1,369 @@
...
+                unexpected_diffs1 = {uu: diff[uu] for uu in list(unexpected_diffs)}
+                raise ValueError(
+                    'Unexpected attribute differences between ensemble members:',
+                    unexpected_diffs1
+                )
+
+            diff_keys = list(diff['values_changed'].keys())
+            all_diff_keys = all_diff_keys | set([ss.replace('root', '') for ss in diff_keys])
+
+        # TODO(JLM): What is this doing? Comment.
```
Comment (jmills-ncar) [inline wrfhydropy/core/ensemble.py]: haha, didn't you write it?

File: wrfhydropy/core/ensemble.py [hunk 3]
```diff
@@ -0,0 +1,369 @@
...
+            self.__member_diffs[dd] = [get_path(dictify(mm), dd) for mm in self.members]
+
+        return(self.__member_diffs)
+
+    def set_member_diffs(
+        self,
+        att_tuple: tuple,
+        values: list
+    ):
+
```
Comment (jmills-ncar) [inline wrfhydropy/core/ensemble.py]: docstring please

File: wrfhydropy/core/ensemble.py [hunk 4]
```diff
@@ -0,0 +1,369 @@
...
+        # their relative dirs, if requested
+        if rm_members_from_memory:
+            run_dirs = [mm.run_dir for mm in self.members]
+            self.members = run_dirs
+
+    def run(
+        self,
+        n_concurrent: int=1
+    ):
+
```
Comment (jmills-ncar) [inline wrfhydropy/core/ensemble.py]: docstring please

File: wrfhydropy/core/ensemble_tools.py [hunk 5]
```diff
@@ -0,0 +1,84 @@
+from boltons.iterutils import remap
+import copy
+from deepdiff.diff import DeepDiff
+
+
+def is_sub_obj(obj):
+    """Test if an object is has a __dict__ (may not be the best definition of an object,
```
Comment (jmills-ncar) [inline wrfhydropy/core/ensemble_tools.py]: include args in docstring

**Comment by jmccreight** [comment]
TRUE DAT

File: wrfhydropy/core/ensemble.py [hunk 6]
```diff
@@ -0,0 +1,369 @@
...
+from .ensemble_tools import DeepDiffEq, dictify, get_sub_objs
+from .job import Job
+from .schedulers import Scheduler
+from .simulation import Simulation
+
+#from .job_tools import solve_model_start_end_times
+
+
+def parallel_compose_addjobs(arg_dict):
+    for jj in arg_dict['jobs']:
```
[the remark this answers is not in the document]
Comment (jmccreight) [reply wrfhydropy/core/ensemble.py]: I think we could debate this. I like double letters for iterands where the letters reference the thing being iterated. It makes clear what in a loop is changing.

File: wrfhydropy/core/ensemble.py [hunk 7]
```diff
@@ -0,0 +1,369 @@
...
+        arg_dict['member'].add(jj)
+    return arg_dict['member']
+
+
+def parallel_compose_addscheduler(arg_dict):
+
```
[the remark this answers is not in the document]
Comment (jmccreight) [reply wrfhydropy/core/ensemble.py]: OK. My pep8 dosent complain about this...

File: wrfhydropy/core/ensemble.py [hunk 8]
```diff
@@ -0,0 +1,369 @@
...
+    # 2) member_dir
+    # # Removed these two until it's obvious we need them
+    # # 3) description
+    # # 4) forcing_source_dir
+
+    def add(
+        self,
+        obj: Union[list, Scheduler, Job]
+    ):
+        """Add an approparite object to an EnsembleSimulation, such as a Simulation, Job, or
```
[the remark this answers is not in the document]
Comment (jmccreight) [reply wrfhydropy/core/ensemble.py]: sorry, stole this from simulation! :) Will update both

**Comment by jmccreight** [comment]
I addressed the previous issues and added some tests. I dont think the coveralls complaint about simulation coverage doing down is correct. the coverage did tick up about 5% with this.

**Pull request merged**